File: sdks/elixir/lib/autonoma/fingerprint.ex
defmodule Autonoma.Fingerprint do
  @moduledoc """
  Deterministic SHA256-based fingerprinting of scenario definitions.
  Produces a 16-character hex string, order-independent for object keys.
  """

  @doc """
  Compute a 16-char hex fingerprint of any JSON-serializable value.
  """
  def compute(value) do
    json = value |> sort_keys() |> Jason.encode!()

    :crypto.hash(:sha256, json)
    |> Base.encode16(case: :lower)
    |> binary_part(0, 16)
  end

  defp sort_keys(value) when is_map(value) do
    value
    |> Enum.sort_by(fn {k, _v} -> k end)
    |> Enum.map(fn {k, v} -> {k, sort_keys(v)} end)
    |> Jason.OrderedObject.new()
  end

  defp sort_keys(value) when is_list(value) do
    Enum.map(value, &sort_keys/1)
  end

  defp sort_keys(value), do: value
end

File: sdks/go/autonoma/fingerprint.go
package autonoma

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"sort"
)

// Fingerprint computes a stable 16-char hex fingerprint of a value.
// Uses SHA256 of JSON-serialized data with sorted object keys.
func Fingerprint(value any) string {
	sorted := sortKeys(value)
	data, _ := json.Marshal(sorted)
	hash := sha256.Sum256(data)
	return hex.EncodeToString(hash[:])[:16]
}

// sortKeys recursively sorts object keys for deterministic serialization.
func sortKeys(value any) any {
	switch v := value.(type) {
	case map[string]any:
		sorted := make(sortedMap, 0, len(v))
		for key, val := range v {
			sorted = append(sorted, sortedMapEntry{Key: key, Value: sortKeys(val)})
		}
		sort.Slice(sorted, func(i, j int) bool {
			return sorted[i].Key < sorted[j].Key
		})
		return sorted
	case []any:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = sortKeys(item)
		}
		return result
	default:
		return value
	}
}

// sortedMap is a slice of key-value pairs that serializes with keys in order.
type sortedMap []sortedMapEntry

type sortedMapEntry struct {
	Key   string
	Value any
}

func (s sortedMap) MarshalJSON() ([]byte, error) {
	buf := []byte{'{'}
	for i, entry := range s {
		if i > 0 {
			buf = append(buf, ',')
		}
		keyBytes, err := json.Marshal(entry.Key)
		if err != nil {
			return nil, err
		}
		valBytes, err := json.Marshal(entry.Value)
		if err != nil {
			return nil, err
		}
		buf = append(buf, keyBytes...)
		buf = append(buf, ':')
		buf = append(buf, valBytes...)
	}
	buf = append(buf, '}')
	return buf, nil
}
